### ISSUE TITLE
CNN (Convolutional Neural Network)

### ISSUE THREAD

**Issue by iammiori** [iammiori/AI_19_SS_CLASS#5]
## 1. Fully Connected (FC) layer 의 문제점

> 1-dimension 데이터만 입력받을 수 있다
> ex) 32x32x3 image -> stretch to 3072x1
=> 말인 즉슨, 3-dimension 데이터를 **flat 하게 해서 입력**해야한다.
= > 결국 3-dimension 데이터의 **공간적 정보가 소실**되는 문제가 발생한다

## 1. 해결 = > CONV layer

> shape 항상 유지, 공간적 정보 유지 가능

**Comment by iammiori** [comment]
## 2. Convolution (=합성곱)

> **Kernel Size** 
   : convolution의 view 결정 
   : (InputSize - KernelSize)/stride + 1 = outputsize 

> **Stride**
   : 이미지 위를 움직일 때의 kernel의 step size. 기본값은 1
   : Stride가 클수록 overlapping 이 덜하고 output 크기가 작아짐

> **Activation Map**
    : 해당 receptive field 에 filter에서 detect 하고자 하는 feature 가 있는지 없는지 알려줌
    : Activation Map Size = (Input Size - Filter Size)/stride + 1

> ex : filter 크기계산법 (stride=1이라 가정)
    : 1) Input 7* 7 => output 5*5
    : sol) 7-x + 1 = 5 따라서, x = 3

> **Padding**
   : sample 테두리 조절해줌
   : 결과값이 작아지는 거 방지
   : 입력값 주위로 0을 넣어서 입력값 크기 키워줌
   : overfitting 도 막아줌
   : output dimension 유지해줌

> **Zero padding 생각해준다면**
    : Output Size = (Input Size + 2*Padding - FilterSize)/stride + 1
    : Input Size + 2*padding = padding 이 있을 때 반영된 Input 크기
    :  zero padding 값은 (stride가 1일때) (Filter Size -1) /2 로 결정 
    : 왜? => 양쪽에 0들을 둬야하니까
    : 1 더하는 이유 : Conv 하기 전, 첫번째 filter가 input과 수행한 횟수
    


## 2-1) code

**input**
`X_img = tf.reshape(X, [-1, 28, 28, 1])`
[입력데이터수,세로,가로,채널수]

**filter**
: detect 하고자 하는 feature에 대한 내용이 담긴 사각형
: Kernel, weight
`W1 = tf.Variable(tf.random_normal([3, 3, 1, 32], stddev=0.01))`
[세로,가로,채널수,필터수]

**output**
`X_img = tf.reshape(X, [-1, 28, 28, 1])`
[출력데이터수,세로,가로,채널수]

**Comment by iammiori** [comment]
## 3. 과정

> input data
=> **first layer** : conv -> relu -> maxpool
=> **second layer** : conv -> relu -> maxpool
=> reshape 
=> fully connected
=> outputdata

## 3-1) conv2d

> filter를 특징으로 뽑아주는 convolution layer
`tf.nn.conv2d(input, filter, strides, padding, use_cudnn_on_gpu=None, data_format=None, name=None)`

**input**
**filter** 
**strides** : 크기가 4인 1차원 list [1,n,n,1] 꼴
**padding** : SAME (output 크기 = input크기) / VALID

**Comment by iammiori** [comment]
### ReLU

**Activation Layer**

> - sigmoid : 미분값이 작아서, 계속 곱하다 보면 0에 수렴
> - tanh
> - ReLU ( Rectified Linear Unit) 

**ReLU**

> - 느린 계산 수정
> - Vanishing Gradient Problem 해결
> - max([x,0)]
> - x<0 -> f(x) = 0 // x>=0 -> f(x)=x

**Comment by iammiori** [comment]
### Pooling
> stride=1 => 겹치는 부분이 많음 => 그중 best match selection 만 쓰자 => overfitting 방지

**Max Pooling**

> - 한 subregion에서 Max 선택
> - Max 쓰는 이유는 feature 잘 표현 해줘서
